### ISSUE TITLE
sanctuary-def dependency

### ISSUE THREAD

**Issue by davidchambers** [sanctuary-js/sanctuary#327]
Sanctuary depends on sanctuary-def, as specified in __package.json__. Furthermore, Sanctuary is only *compatible* with certain versions of sanctuary-def. @codedmart recently brought it to my attention on Gitter that the example for [`create`][1] does not work if one uses the current version of Sanctuary (v0.11.1) with the current version of sanctuary-def (v0.9.0).

There are (at least) two things we should consider:

  - specifying this constraint in __package.json__, if possible; and
  - versioning the two packages such that `sanctuary@X.Y.Z` always depends on and works with `sanctuary-def@X.Y.Z`.

[1]: https://github.com/sanctuary-js/sanctuary/tree/v0.11.1#create

**Comment by davidchambers** [comment]
Could we use [`peerDependencies`][1] to express the constraint?

```json
  "dependencies": {
    "sanctuary-def": "0.9.0",
    "sanctuary-type-classes": "2.0.1",
    "sanctuary-type-identifiers": "1.0.0"
  },
  "peerDependencies": {
    "sanctuary-def": "0.9.0"
  }
```


[1]: https://docs.npmjs.com/files/package.json#peerdependencies

**Comment by Avaq** [comment]
> versioning the two packages such that `sanctuary@X.Y.Z` always depends on and works with `sanctuary-def@X.Y.Z`



> Could we use `peerDependencies` to express the constraint?

I've never seen `peerDependencies` used like that, though I haven't come across the situation where a package is both a *dependency* and a *plugin* to another package. I guess it might win us an automatic warning when the user tries to use an incompatible `def`, but I think it's all a bit messy. It relates my the concerns I expressed [here](https://github.com/Avaq/Fluture/issues/63#issuecomment-275618345).

The problem, I think, is that multiple parties are creating black-box data structures (type definitions) using `sanctuary-def` and multiple parties need to interpret (call `def` on) those structures using `sanctuary-def`. Maybe it's possible for the future to ship the interpreter with "batteries included", eg:

```js
const sanctuary = require('sanctuary');
const $ = require('sanctuary-def');
const env = $.env.concat(sanctuary.env).concat(myEnv);
const def = TYPE_CHECKING ? $.createTypecheckedDef(env) : $.def;

const S = sanctuary.create(def);
```

`sanctuary` would have to have a Peer Dependency on `sanctuary-def`, but *not* direct dependency. It's compatibility with `def` would only change if the type signature of `def()` changes, but not when the internal data structures change. Maybe It's a bad idea. Might be one for **Ideas about env** ;)

**Comment by Avaq** [comment]
Ah I just realized the flaw in my idea above: In order to provide an `env`, we would still need `sanctuary-def`.

**Comment by Avaq** [comment]
But it *does* seem to me like removing `sanctuary-def` from dependencies and keeping it as only a peer-dependency would reduce complexity.

**Comment by davidchambers** [comment]
> But it *does* seem to me like removing `sanctuary-def` from dependencies and keeping it as only a peer-dependency would reduce complexity.

I agree. `const S = sanctuary.create(def);` seems very nice to me! Let's seriously consider this change once we get v0.12.0 out the door. :)

**Comment by Avaq** [comment]
I think this is what we should do:

1. Separate all of sanctuary into small modules, each with a *peer* dependency on `sanctuary-def`:
   * `sanctuary-either` to contain the Either type, its Repr and associated functions
   * `sanctuary-maybe` to contain the Maybe type, its Repr and associated functions
   * `sanctuary-list` for list related functions
   * `sanctuary-string` for string related functions
   * `sanctuary-strmap` for the StrRep type and related functions
   * ...more
1. Create a `sanctuary-prelude` library which takes from each of the above libs to provide a set of common tools.
1. The `sanctuary` library will have a *single* function. To load `sanctuary-prelude` and `sanctuary-def` and instantiate the prelude it with a default `def`.

People who want an out-of-the-box experience will use `sanctuary`, people who want a customized one will use `sanctuary-prelude`, people who want only parts of the functionality will assemble their own utility belt from the available underlying libs.

This tackles several problems:

1. "The library is monolithic": Now we can choose what we want to include. Maybe in the future, each lib can even slit up into stuff like `sanctuary-maybe-type`, `sanctuary-maybe-from-nullable`.
2. (_and this is personal_) "my colleagues are using `require('sanctuary')` directly, which defeats the point of disabling type-checking": Now, I will depend only on `sanctuary-prelude`, and provide my own assembly which they *have* to use.
3. "my sanctuary-def is not compatible with yours": Since all packages but `sanctuary` have it as a peer dependency, they will all use the same version and NPM will show it when one is not compatible and should be down or upgraded.

**Comment by rjmk** [comment]
Is it `UnaryType` that breaks things currently?

**Comment by rickmed** [comment]
I really don't know the internals of sanctuary so I have no idea what is possible but I assume that eacho of those packages would be sanctuary-def (78 kb) and sanctuary-type-classes (45kb). Right now sanctuary is 100kb, let's say sanctuary-maybe would me 15kb. Would it make sense to have a separate package of 15kb with 123kb of dependencies. In that case, me as a user, would prefer to download a es6 modularized sanctuary and tree shake it myself. Other option is to ask if you are downloading a single package probably you are working on a small project and you would not need type checking, so remove type checking on individual packages -if possible.
